### ISSUE TITLE
💅 New false positive for extra non null assertion in v2.3.9

### ISSUE THREAD

**Issue by aaronjordan** [biomejs/biome#8475]
### Environment information

```bash
CLI:
  Version:                      2.3.9
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-ghostty
  JS_RUNTIME_VERSION:           v24.11.1
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         pnpm/10.18.2

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.json
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false
  HTML full support enabled:    unset

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  true
  Enabled rules:
    a11y/noAccessKey
    a11y/noAriaHiddenOnFocusable
    a11y/noAriaUnsupportedElements
    a11y/noAutofocus
    a11y/noDistractingElements
    a11y/noHeaderScope
    a11y/noInteractiveElementToNoninteractiveRole
    a11y/noLabelWithoutControl
    a11y/noNoninteractiveElementToInteractiveRole
    a11y/noNoninteractiveTabindex
    a11y/noPositiveTabindex
    a11y/noRedundantAlt
    a11y/noRedundantRoles
    a11y/noStaticElementInteractions
    a11y/noSvgWithoutTitle
    a11y/useAltText
    a11y/useAnchorContent
    a11y/useAriaActivedescendantWithTabindex
    a11y/useAriaPropsForRole
    a11y/useAriaPropsSupportedByRole
    a11y/useButtonType
    a11y/useFocusableInteractive
    a11y/useGenericFontNames
    a11y/useHeadingContent
    a11y/useHtmlLang
    a11y/useIframeTitle
    a11y/useKeyWithClickEvents
    a11y/useKeyWithMouseEvents
    a11y/useMediaCaption
    a11y/useSemanticElements
    a11y/useValidAnchor
    a11y/useValidAriaProps
    a11y/useValidAriaRole
    a11y/useValidAriaValues
    a11y/useValidAutocomplete
    a11y/useValidLang
    complexity/noAdjacentSpacesInRegex
    complexity/noArguments
    complexity/noBannedTypes
    complexity/noCommaOperator
    complexity/noEmptyTypeParameters
    complexity/noExtraBooleanCast
    complexity/noFlatMapIdentity
    complexity/noImportantStyles
    complexity/noStaticOnlyClass
    complexity/noThisInStatic
    complexity/noUselessCatch
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessStringRaw
    complexity/noUselessSwitchCase
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noUselessUndefinedInitialization
    complexity/useArrowFunction
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useIndexOf
    complexity/useLiteralKeys
    complexity/useNumericLiterals
    complexity/useOptionalChain
    complexity/useRegexLiterals
    complexity/useSimpleNumberKeys
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noStringCaseMismatch
    correctness/noSwitchDeclarations
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/useGraphqlNamedOperations
    correctness/useIsNan
    correctness/useParseIntRadix
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    performance/noAccumulatingSpread
    performance/noDynamicNamespaceImportAccess
    security/noBlankTarget
    security/noGlobalEval
    style/noDescendingSpecificity
    style/noNonNullAssertion
    style/useArrayLiterals
    style/useConst
    style/useDeprecatedReason
    style/useExponentiationOperator
    style/useExportType
    style/useImportType
    style/useLiteralEnumMembers
    style/useNodejsImportProtocol
    style/useShorthandFunctionType
    style/useTemplate
    suspicious/noApproximativeNumericConstant
    suspicious/noAssignInExpressions
    suspicious/noAsyncPromiseExecutor
    suspicious/noBiomeFirstException
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConfusingVoidType
    suspicious/noConstEnum
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDocumentCookie
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noEmptyBlock
    suspicious/noEmptyInterface
    suspicious/noExplicitAny
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noGlobalIsNan
    suspicious/noImplicitAnyLet
    suspicious/noImportAssign
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noNonNullAssertedOptionalChain
    suspicious/noOctalEscape
    suspicious/noPrototypeBuiltins
    suspicious/noQuickfixBiome
    suspicious/noRedeclare
    suspicious/noRedundantUseStrict
    suspicious/noSelfCompare
    suspicious/noShadowRestrictedNames
    suspicious/noShorthandPropertyOverrides
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noTemplateCurlyInString
    suspicious/noThenProperty
    suspicious/noTsIgnore
    suspicious/noUnknownAtRules
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noUselessEscapeInString
    suspicious/noUselessRegexBackrefs
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useBiomeIgnoreFolder
    suspicious/useDefaultSwitchClauseLast
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useIsArray
    suspicious/useIterableCallbackReturn
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0
```

### Rule name

lint/suspicious/noExtraNonNullAssertion

### Playground link

https://biomejs.dev/playground/?code=LwAvACAAYgBpAG8AbQBlAC0AaQBnAG4AbwByAGUALQBhAGwAbAAgAGwAaQBuAHQALwBzAHQAeQBsAGUALwBuAG8ATgBvAG4ATgB1AGwAbABBAHMAcwBlAHIAdABpAG8AbgA6ACAAcwBlAHQAIABpAG4AIAByAGUAYQBsACAAcAByAG8AagBlAGMAdAAgAGYAaQBsAGUACgAKAGMAbwBuAHMAdAAgAGIAbABvAGIAOgAgAFIAZQBjAG8AcgBkADwAcwB0AHIAaQBuAGcALAAgAFIAZQBjAG8AcgBkADwAcwB0AHIAaQBuAGcALAAgAHMAdAByAGkAbgBnAD4APgAgAD0AIAB7AAoACQBhADoAIAB7AH0ALAAKAAkAYgA6ACAAewAgAGMAOgAgACIAaABlAGwAbABvACIALAAgAGQAOgAgACIAdwBvAHIAbABkACEAIgAgAH0ALAAKAH0AOwAKAAoALwAvACAAcwBhAGYAZQAgAGEAYwBjAGUAcwBzACAAdABvACAAYgBsAG8AYgAgAGEAcwAgAGUAeABwAGUAYwB0AGUAZAAKAGMAbwBuAHMAdAAgAF8AZAAxACAAPQAgAGIAbABvAGIALgBiAD8ALgBkAD8ALgBsAGUAbgBnAHQAaAA7AAoACgAvAC8AIAB1AG4AcwBhAGYAZQAgAGEAYwBjAGUAcwBzACAAdABvACAAYgBsAG8AYgAgAHMAaABvAHUAbABkACAAcABlAHIAbQBpAHQAIABiAG8AdABoACAAJwAhACcACgBjAG8AbgBzAHQAIABfAGQAMgAgAD0AIABiAGwAbwBiAC4AYgAhAC4AZAAhAC4AbABlAG4AZwB0AGgAOwA%3D

### Expected result

Biome v2.3.8 and previous would permit an expression with chained non-null assertions, but v2.3.9 is marking these assertions as extras.

```ts
// from playground:
const _d2 = blob.b!.d!.length;
```

Applying the safe fix from Biome in this case will cause TypeScript to fail when the `noUncheckedIndexedAccess` flag is set in `tsconfig.json`.

```
> tsc --noEmit

test.ts:12:13 - error TS18048: 'blob.b' is possibly 'undefined'.

12 const _d2 = blob.b.d!.length;
               ~~~~~~
```

For reference, here are the config files I'm using locally:

<details><summary>biome.json</summary>

```json
{
    "$schema": "https://biomejs.dev/schemas/2.3.9/schema.json",
    "linter": {
        "enabled": true,
        "rules": {
            "recommended": true
        }
    }
}
```

</details>

<details><summary>package.json</summary>

```json
{
    "name": "biome-bug",
    "version": "1.0.0",
    "description": "",
    "main": "test.ts",
    "scripts": {
        "lint": "biome check test.ts",
        "typecheck": "tsc --noEmit"
    },
    "keywords": [],
    "author": "",
    "license": "ISC",
    "packageManager": "pnpm@10.18.2",
    "dependencies": {
        "@biomejs/biome": "2.3.9",
        "typescript": "^5.9.3"
    }
}
```

</details>

<details><summary>tsconfig.json</summary>

```json
{
    "compilerOptions": {
        "target": "ESNext",
        "strict": true,
        "noUncheckedIndexedAccess": true,
        "skipLibCheck": true
    },
    "include": ["test.ts"]
}
```

</details>

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by jakeleventhal** [comment]
Seeing this as well

**Comment by BenjaminBrienen** [comment]
https://github.com/wgsl-analyzer/wgsl-analyzer/pull/771
